### DISCUSSION TITLE
"Intersection results could not be merged" does not mention the cause

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [colinhacks/zod#3194] [Q&A] [answered]
Zod is great and I'm enjoying the features around intersecting/merging objects. But for `.and`, when the intersections results can't be merged for some reason, the error message thrown at runtime doesn't provide any details.

> An error occurred.
> 
> ```plaintext
> Intersection results could not be merged
> ```
> 
> Stack Trace
> 
> ```plaintext
> ZodError: [
>   {
>     "code": "invalid_intersection_types",
>     "path": [],
>     "message": "Intersection results could not be merged"
>   }
> ]
>     at get error [as error] (file:///Users/josh/repos/dot-com/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/index.mjs:538:31)
>     at ZodIntersection.parse (file:///Users/josh/repos/dot-com/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/index.mjs:638:22)
>     at /Users/josh/repos/dot-com/src/pages/rss.xml.ts:26:48
>     at Array.map (<anonymous>)

Not sure if this is a bug report or feature request, but: could the error message also include which field(s) are problematic?

Note that I'm not trying to report that the error is wrong. The error may be right - it just informative enough to explain why clearly.

Sorry that I'm not familiar enough with Zod to provide a standalone reproduction (I tried the Stackblitz Astro starter but couldn't get past TypeScript errors, amusingly). But here's how to get it on a real world project:

```plaintext
git clone https://github.com/JoshuaKGoldberg/dot-com
cd dot-com
git checkout 708081f # ncu-u-for-vite-5 branch
pnpm i
pnpm dev
open http://localhost:4321/rss.xml
```

**Comment by JacobWeisenburger** [comment, marked as answer]
from what I can see the problem code is this:

```ts
import { rssSchema } from "@astrojs/rss"

const rssSchemaStrict = rssSchema.and(
    z.object( {
        link: z
            .string()
            .transform( ( link ) =>
                link.replace( /^src\/content/, "" ).replace( /index.mdx$/, "" ),
            ),
        pubDate: z.date(),
        title: z.string(),
    } ),
)
```

it seems that `rssSchema` is of type `ZodEffect`, so that means it has a `transform` or something at the end of the schema which makes it not an object schema, which means it can't be used with `.and()`

I suggest you use the schema from here instead:
https://github.com/withastro/astro/blob/main/packages/astro-rss/src/schema.ts

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by JoshuaKGoldberg** [reply]
Thanks, I appreciate the help debugging the userland issue 🙂. But, again, I'm trying to report that Zod is giving a poor error message for this issue. If I (the user) am providing bad inputs to a function then Zod should IMO give a better error message for that.

Can we move this back into an issue, please?

**Reply by JoshuaKGoldberg** [reply]
ping @JacobWeisenburger - am I missing something in thinking that the root "Zod is giving a not-helpful-enough error" issue is still unsolved?

I'm hesitant to file a second issue, but don't mind if that's the right way to get this visible.

**Reply by JacobWeisenburger** [reply]
I made it into an issue again when you asked 3 weeks ago.
https://github.com/colinhacks/zod/issues/3196

**Reply by JoshuaKGoldberg** [reply]
Oh! Sorry for the ping, I'd missed that. GitHub Discussions don't have mentions added automatically the way issues and PRs do. Thanks!
